docs/feature_midi.md

```diff
@@ -0,0 +1,299 @@
+
+# MIDI
+
+## Enabling MIDI
+
+Add the following to the keyboard's rules.mk
+
+```c
+MIDI_ENABLE = yes
+```
+
+## Using Basic MIDI
+
+Enabling Basic MIDI
+
+Add the following to your keyboard's config.h
+
+```c
+#define MIDI_BASIC
+```
+
+With basic MIDI you will only be able to send `NOTE_ON` and `NOTE_OFF` Commands to the internal MIDI device, meaning that keycodes like `MI_OCTU` and `MI_OCTD` will not work.
+
+## Advanced MIDI
+
+Enabling Advanced MIDI
+
+Add the following to your keyboard's config.h
+
+```c
+#define MIDI_ADVANCED
+```
+
+With advanced MIDI you can do things like octave shifts, channel changes, velocity changes, and modulation.
+
+### Sending MIDI Control Codes (MIDI CC)
+
+If you're aiming to emulate the features of something like a Launchpad or other MIDI controller you'll need to access the internal MIDI device directly.
+
+When sending MIDI CC you don't get the advantages of a preimplemented keycode and you will need to implement custom keycodes if you want to use them in your keymap directly using the function `process_record_user`. It is the same process as implementing custom keycodes for macros.
+
+For an overview of that process look at: [Macros](feature_macros.md)
+
+1. First enable MIDI_ADVANCED
+2. use an `extern MidiDevice midi_device;` statement to bring the MIDI device into scope.
+3. Send a control code with the `midi_send_cc(*midi_device, channel, number, value)`
+
+For reference of all the possible control code numbers see [MIDI Specification](#midi-specification)
+
+#### Example code for using Generic On Off Switches as per MIDI Specification.
+```c
+#include QMK_KEYBOARD_H
+
+extern MidiDevice midi_device;
+
+/*
+MIDI CC codes for generic ON/OFF swiches
+80,81,82,83
+values off = 0-63
+*/
+
+#define OFF 0
+
+/*values on = 64-127*/
+
+#define ON 127
+
+enum custom_keycodes { MIDI_CC80 = SAFE_RANGE, MIDI_CC81, MIDI_CC82, MIDI_CC83 };
+
+bool process_record_user(uint16_t keycode, keyrecord_t *record) {
+    switch (keycode) {
+        case MIDI_CC80:
+            if (record->event.pressed) {
+                midi_send_cc(&midi_device, midi_config.channel, 80, ON);
+            } else {
+                midi_send_cc(&midi_device, midi_config.channel, 80, OFF);
+            }
+            return true;
+    }
+    return true;
+};
+
+const uint16_t PROGMEM keymaps[][MATRIX_ROWS][MATRIX_COLS] = {
+    [0] = {
+        {MIDI_CC80, MIDI_CC81, MIDI_CC82, MIDI_CC83},
+        // ...
+    },
+};
+
+```
+
+### Available MIDI Keycodes
+
+| Keycode                    | MIDI Code | Description                                             |
```

Comment:
Keycode values column isn't needed, they are subject to change.

Reply:
Will remove.